### ISSUE TITLE
Seek and read from closed MemoryIO

### ISSUE THREAD

**Issue by braingram** [asdf-format/asdf#1539]
ASDF will read from a 'closed' `generic_io.MemoryIO` (like what is created when reading from an `io.BytesIO` instance).

Running the following:
```python
import io
import asdf
import numpy

b = io.BytesIO()
arrs = [numpy.arange(10) + i for i in range(10)]
af = asdf.AsdfFile({'arrs': arrs})
af.write_to(b)

b.seek(0)
with asdf.open(b) as raf:
    arefs = raf['arrs']

print(b.tell())
print(arefs[1][:])
print(b.tell())
```
produces:
```bash
2902
[ 1  2  3  4  5  6  7  8  9 10]
1830
```
showing that the array loading from the 'closed' `MemoryIO` resulting in a change in the file pointer position.

**Comment by braingram** [comment]
This also affects opening and reading from regular files:
```python
import asdf
import numpy

fn = 'test.asdf'
arrs = [numpy.arange(10) + i for i in range(10)]
af = asdf.AsdfFile({'arrs': arrs})
af.write_to(fn)

fp = open(fn, 'rb')
with asdf.open(fp) as raf:
    arefs = raf['arrs']

print(fp.tell())
print(arefs[1][:])
print(fp.tell())
```

```bash
2902
[ 1  2  3  4  5  6  7  8  9 10]
1830
```

**Comment by braingram** [comment]
Fixed in #1537 both examples above now no longer seek the file pointer and raise an exception:
```
OSError: Attempt to load block from closed file
```
